File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/PositionedPlayingFile.java
package com.lasthopesoftware.bluewater.client.playback.file;

import android.support.annotation.NonNull;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.file.volume.ManagePlayableFileVolume;

public final class PositionedPlayingFile implements Comparable<PositionedPlayingFile> {
	private final PlayingFile playingFile;
	private final ManagePlayableFileVolume playableFileVolumeManager;
	private final PositionedFile positionedFile;

	public PositionedPlayingFile(int playlistPosition, PlayingFile playingFile, ManagePlayableFileVolume playableFileVolumeManager, ServiceFile serviceFile) {
		this(playingFile, playableFileVolumeManager, new PositionedFile(playlistPosition, serviceFile));
	}

	public PositionedPlayingFile(PlayingFile playingFile, ManagePlayableFileVolume playableFileVolumeManager, PositionedFile positionedFile) {
		this.playableFileVolumeManager = playableFileVolumeManager;
		this.positionedFile = positionedFile;
		this.playingFile = playingFile;
	}

	public int getPlaylistPosition() {
		return positionedFile.getPlaylistPosition();
	}

	public ServiceFile getServiceFile() {
		return positionedFile.getServiceFile();
	}

	public PlayingFile getPlayingFile() {
		return playingFile;
	}

	public ManagePlayableFileVolume getPlayableFileVolumeManager() {
		return playableFileVolumeManager;
	}

	@Override
	public int compareTo(@NonNull PositionedPlayingFile other) {
		return positionedFile.compareTo(other.positionedFile);
	}

	@Override
	public boolean equals(Object obj) {
		return obj instanceof PositionedPlayingFile && compareTo((PositionedPlayingFile)obj) == 0;
	}

	@Override
	public int hashCode() {
		return positionedFile.hashCode();
	}

	public PositionedFile asPositionedFile() {
		return positionedFile;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/PositionedPlayingFile.kt
package com.lasthopesoftware.bluewater.client.playback.file

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile
import com.lasthopesoftware.bluewater.client.playback.file.progress.ReadFileProgress
import com.lasthopesoftware.bluewater.client.playback.file.volume.ManagePlayableFileVolume
import com.namehillsoftware.handoff.promises.Promise
import org.joda.time.Duration

class PositionedPlayingFile(
	val playingFile: PlayingFile,
	val playableFileVolumeManager: ManagePlayableFileVolume,
	private val positionedFile: PositionedFile
) : Comparable<PositionedPlayingFile>, ReadFileProgress {
	constructor(playlistPosition: Int, playingFile: PlayingFile, playableFileVolumeManager: ManagePlayableFileVolume, serviceFile: ServiceFile)
		: this(playingFile, playableFileVolumeManager, PositionedFile(playlistPosition, serviceFile))

	val playlistPosition: Int
		get() = positionedFile.playlistPosition
	val serviceFile: ServiceFile
		get() = positionedFile.serviceFile
	override val progress: Promise<Duration>
		get() = playingFile.promisePlayedFile().progress

	override fun compareTo(other: PositionedPlayingFile): Int = positionedFile.compareTo(other.positionedFile)

	override fun equals(other: Any?): Boolean =
		(other as? PositionedPlayingFile)?.let { compareTo(it) == 0 } ?: false

	override fun hashCode(): Int = positionedFile.hashCode()

	fun asPositionedFile(): PositionedFile = positionedFile
}
